### PR TITLE
SOR V3 - Use Balancer Maths for V3 Liquidity

### LINKED ISSUE #602: SOR V3 - Use Balancer Maths for V3 liquidity
Current pool maths follow Balancer V2 structure.

This task aims to add V3 pool math structure to the SOR, which provides a set of base pool math for add/remove liquidity and requires simply the invariant and "reverse invariant" in order to support new pool types.

It should initially support Weighted and Stable pools, and should be easily extensible to support adding new pool types.

### PR DESCRIPTION
**Pull request by brunoguerios**
Closes #602

### CODE REVIEW

File: modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts
Comment (johngrantuk) [inline modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: The best way to double check this is to run a test with token that has < 18 decimals.

File: modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts
Comment (johngrantuk) [inline modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: Do you think its useful to extend from `ComposableStablePool`? Feels like its mixing things up a bit and doesn't really offer any benefit.

Comment (brunoguerios) [reply modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: I use it to inherit the constructor/fromPrismaPool methods, so I get a BasePool, which is what the SOR expects as input. The alternative would be to copy/paste them here, but I think extending makes a bit more sense.
What do you think?

File: modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts
Comment (brunoguerios) [reply modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: True! I'll do that over here 👍

File: modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts
Comment (johngrantuk) [reply modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: It would probably be better to separate out the shared methods and use them where needed. Otherwise StablePoolV3 ends up inheriting a bunch of other stuff that isn't needed and may end up causing confusion elsewhere.

Comment (brunoguerios) [reply modules/sor/sorV2/lib/pools/composableStable/stablePoolV3.ts]: I agree it may be confusing to inherit things that are not being used. I'll reimplement as separate pools. And I'll do the same with Weighted Pool V3, so it's easier to follow when we're maintaining the code.

File: modules/sor/sorV2/lib/pools/v2/composableStable/composableStablePool.ts
Comment (brunoguerios) [inline modules/sor/sorV2/lib/pools/v2/composableStable/composableStablePool.ts:322]: This applies the same maxSwapAmount logic implemented on Balancer Maths

File: modules/sor/sorV2/lib/pools/v2/composableStable/composableStablePool.ts
Comment (brunoguerios) [inline modules/sor/sorV2/lib/pools/v2/composableStable/composableStablePool.ts:324]: This applies the same maxSwapAmount logic implemented on Balancer Maths

File: modules/sor/sorV2/lib/pools/v2/fx/fxPool.ts
Comment (brunoguerios) [inline modules/sor/sorV2/lib/pools/v2/fx/fxPool.ts:1]: Reorganized imports on a few SOR files and added absolute paths to make it easier to read.

File: modules/sor/sorV2/lib/pools/v2/weighted/weightedPool.ts
Comment (brunoguerios) [inline modules/sor/sorV2/lib/pools/v2/weighted/weightedPool.ts:11]: Extracted WeightedPoolToken to its own file. Felt it would be better like this, because WeightedPoolV3 reuses it.

File: modules/sor/sorV2/lib/static.ts
Comment (brunoguerios) [inline modules/sor/sorV2/lib/static.ts]: @franzns - v3 stable pools actually have their type as 'STABLE', right? I'll update this.

File: package.json
Comment (brunoguerios) [inline package.json]: Required for using absolute paths on test files.

File: tsconfig.json
Comment (brunoguerios) [inline tsconfig.json]: Adds absolute paths

File: modules/sor/sorV2/lib/static.ts
Comment (franzns) [reply modules/sor/sorV2/lib/static.ts]: Correct.

**Review by johngrantuk — APPROVED** [review]
Assuming testing has been done locally and all is ok it all LGTM.

**Comment by franzns** [comment]
@brunoguerios Can you please add a changeset? After that you can merge :)

**Comment by brunoguerios** [comment]
I still want to add a few tests to make sure the code is working as expected. After that I'll add changeset and merge 👍

**Comment by brunoguerios** [comment]
FYI - I broke something when reorganizing files/folders to include v3 pools within the SOR and unit tests started failing. I couldn't find the issue, but I found a commit where it stops working, so I'll revert to that and reimplement. I should get it ready to go by the end of the day.

**Comment by brunoguerios** [comment]
I finally found the issue and it's really odd 👀 
Apparently import orders matter - here's the change that broke tests 🙄 
```
import { WeightedPool } from './weightedPool';
import {
    prismaPoolDynamicDataFactory,
    prismaPoolFactory,
    prismaPoolTokenDynamicDataFactory,
    prismaPoolTokenFactory,
} from '../../../../../../../test/factories';
```
to:
```
import {
    prismaPoolDynamicDataFactory,
    prismaPoolFactory,
    prismaPoolTokenDynamicDataFactory,
    prismaPoolTokenFactory,
} from '../../../../../../../test/factories';
import { WeightedPool } from './weightedPool';
```
Is that known to any of you guys? I'll try to do a bit more research on this, but for now I'll focus on finishing the PR.

**Comment by brunoguerios** [comment]
Closing this in favor of #647

**Pull request closed without merging**
